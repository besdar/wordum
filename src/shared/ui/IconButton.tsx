import React, {useState} from 'react';
import {IconButton as PaperButton} from 'react-native-paper';

type Props = React.ComponentProps<typeof PaperButton> & {
  noLoading?: boolean;
};

export const IconButton = ({
  onPress,
  disabled,
  loading,
  noLoading,
  ...props
}: Props) => {
  const [isLoading, setIsLoading] = useState(false);

  return (
    <PaperButton
      {...props}
      onPress={e => {
        setIsLoading(true);
        Promise.resolve(onPress?.(e)).finally(() => setIsLoading(false));
      }}
      loading={!noLoading && (loading || isLoading)}
      disabled={disabled || isLoading}
    />
  );
};
