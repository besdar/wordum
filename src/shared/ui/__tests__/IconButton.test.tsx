import React from 'react';
import {render, screen, userEvent} from '@testing-library/react-native';
import {IconButton} from '../IconButton';
import {View} from 'react-native';

describe('IconButton', () => {
  it('should match the snapshot', () => {
    const {toJSON} = render(<IconButton icon={() => <View />} />);

    expect(toJSON()).toMatchSnapshot();
  });

  it('should not call onPress when disabled', async () => {
    const mockOnPress = jest.fn();

    render(<IconButton onPress={mockOnPress} disabled icon={() => <View />} />);

    const button = screen.getByRole('button');

    const user = userEvent.setup();
    await user.press(button);

    expect(mockOnPress).not.toHaveBeenCalled();
  });
});
