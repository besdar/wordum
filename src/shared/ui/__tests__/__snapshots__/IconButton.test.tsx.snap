// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton should match the snapshot 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "transparent",
      "borderRadius": 20,
      "height": 40,
      "margin": 6,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 0,
        "width": 0,
      },
      "shadowOpacity": 0,
      "shadowRadius": 0,
      "width": 40,
    }
  }
  testID="icon-button-container-outer-layer"
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "transparent",
        "borderColor": "rgba(121, 116, 126, 1)",
        "borderRadius": 20,
        "borderWidth": 0,
        "elevation": 0,
        "flex": 1,
        "overflow": "hidden",
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 0,
          "width": 0,
        },
        "shadowOpacity": 0,
        "shadowRadius": 0,
      }
    }
    testID="icon-button-container"
  >
    <View
      accessibilityComponentType="button"
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityTraits="button"
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      centered={true}
      collapsable={false}
      focusable={true}
      hitSlop={
        {
          "bottom": 6,
          "left": 6,
          "right": 6,
          "top": 6,
        }
      }
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "overflow": "hidden",
          },
          [
            {
              "alignItems": "center",
              "flexGrow": 1,
              "justifyContent": "center",
            },
            {
              "borderRadius": 20,
            },
          ],
        ]
      }
      testID="icon-button"
    >
      <View />
    </View>
  </View>
</View>
`;
